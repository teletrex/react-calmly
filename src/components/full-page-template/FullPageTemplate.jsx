/* <LICENSE>
*
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
*
* </LICENSE>
*  */



import React, { useState, useMemo, useCallback } from 'react';
import settings  from '../../settings';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import isFunction from 'lodash/isFunction';

import { H2 } from '../heading';

import FullPageTemplateContext from './FullPageTemplateContext';

const { prefix } = settings;

const wrapElement = (headerProp, Component) =>
  typeof headerProp === 'string' ? <Component>{headerProp}</Component> : headerProp;

const FullPageTemplate = ({
  children,
  header,
  subheader,
  actionBar,
  className,
  scrollableTable,
  actionBarTitleAlign,
}) => {
  const defaultClassName = `${prefix}--full-page-template`;
  const isSubheaderVisible = !!subheader;
  const subheaderClassNames = classNames(`${defaultClassName}__header-right`, {
    [`${defaultClassName}__header-right__marginless`]: !isSubheaderVisible,
    [`${defaultClassName}__header-right__title-aligned`]: actionBarTitleAlign,
  });
  const [tableRef, setTableRef] = useState(null);
  const [shouldShowTable, setShowTable] = useState(false);

  const onRef = useCallback(value => {
    setTableRef(value);
  }, []);
  const contextValue = useMemo(
    () => ({
      tableWrapperRef: tableRef,
      registerTable: () => setShowTable(true),
      unregisterTable: () => setShowTable(false),
    }),
    [tableRef]
  );

  return (
    <FullPageTemplateContext.Provider value={contextValue}>
      <main className={classNames(defaultClassName, className)}>
        <div className={`${defaultClassName}__heading`}>
          <div className={`${defaultClassName}__header-left`}>
            {header && (
              <div className={`${defaultClassName}__header-wrapper`}>{wrapElement(header, H2)}</div>
            )}
            {subheader && (
              <div className={`${defaultClassName}__subheader-wrapper`}>
                <div className={`${defaultClassName}__subheader-left`}>
                  <p>{isFunction(subheader) ? subheader() : subheader}</p>
                </div>
              </div>
            )}
          </div>
          {actionBar && <div className={subheaderClassNames}>{actionBar}</div>}
        </div>
        <div className={`${defaultClassName}__content`}>
          {children}
          {(scrollableTable || shouldShowTable) && (
            <div
              ref={onRef}
              className={classNames(`${defaultClassName}__scrollable-table`, className)}
            >
              {scrollableTable}
            </div>
          )}
        </div>
      </main>
    </FullPageTemplateContext.Provider>
  );
};

FullPageTemplate.propTypes = {
  actionBar: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]),
  actionBarTitleAlign: PropTypes.bool,
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]).isRequired,
  className: PropTypes.string,
  header: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
    PropTypes.string,
  ]),
  scrollableTable: PropTypes.node,
  subheader: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
    PropTypes.string,
  ]),
};

FullPageTemplate.defaultProps = {
  actionBar: null,
  actionBarTitleAlign: false,
  className: '',
  header: '',
  scrollableTable: null,
  subheader: null,
};

export { FullPageTemplate };
