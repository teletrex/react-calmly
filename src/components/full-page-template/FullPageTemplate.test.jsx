/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */



import React, { useContext } from 'react';
import { render } from '@testing-library/react';
import { mount } from 'enzyme';
import { renderToJson } from 'enzyme-to-json';
import { Header } from 'carbon-components-react';

import FullPageTemplate, { FullPageTemplateContext } from '.';

describe('FullPageTemplate', () => {
  it('verify if created', () => {
    const wrapper = mount(<FullPageTemplate>Content</FullPageTemplate>);
    expect(wrapper.find(FullPageTemplate)).toHaveLength(1);
  });

  it('with scrollable table', () => {
    const wrapper = mount(
      <FullPageTemplate scrollableTable={<div>scrollable table</div>}>Content</FullPageTemplate>
    );

    expect(renderToJson(wrapper.render())).toMatchSnapshot();
  });

  it('should render wrapper for table when table is registered', () => {
    const TableComponent = () => {
      const { registerTable } = useContext(FullPageTemplateContext);

      registerTable();

      return null;
    };

    const wrapper = mount(
      <FullPageTemplate>
        <div>Content</div>
        <TableComponent>table</TableComponent>
      </FullPageTemplate>
    );

    expect(wrapper.find('.bx--full-page-template__scrollable-table')).toHaveLength(1);
  });

  it('should render header wrapper with string', () => {
    const { getByText } = render(
      <FullPageTemplate header="Test header">
        <div>Content</div>
      </FullPageTemplate>
    );

    expect(getByText('Test header')).toBeInTheDocument();
  });

  it('should render header wrapper with component', () => {
    const { getByLabelText } = render(
      <FullPageTemplate header={<Header aria-label="header">Test header</Header>}>
        <div>Content</div>
      </FullPageTemplate>
    );

    expect(getByLabelText('header')).toBeInTheDocument();
  });

  it('should render subheader as string', () => {
    const { getByText } = render(
      <FullPageTemplate subheader="Subheader test">
        <div>Content</div>
      </FullPageTemplate>
    );

    expect(getByText('Subheader test')).toBeInTheDocument();
  });

  it('should render actionbar', () => {
    const ActionBar = () => <button type="button">Action Button</button>;

    const { getByText } = render(
      <FullPageTemplate actionBar={<ActionBar />}>
        <div>Content</div>
      </FullPageTemplate>
    );

    expect(getByText('Action Button')).toBeInTheDocument();
  });
});
