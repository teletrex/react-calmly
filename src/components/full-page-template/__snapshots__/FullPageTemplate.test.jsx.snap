// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FullPageTemplate with scrollable table 1`] = `
<main
  class="bx--full-page-template"
>
  <div
    class="bx--full-page-template__heading"
  >
    <div
      class="bx--full-page-template__header-left"
    />
  </div>
  <div
    class="bx--full-page-template__content"
  >
    Content
    <div
      class="bx--full-page-template__scrollable-table"
    >
      <div>
        scrollable table
      </div>
    </div>
  </div>
</main>
`;
